README
MasonX-Maypole

Provides a Mason frontend and view to Maypole 2.

Change log
==========

0.2_05 Sat Nov 20, 22:32:00 2004
    - fixed a bug in send_output - wasn't returning if delegating to SUPER

0.2_04 Tue Nov 16, 01:05:00 2004
    - removed debugging messages, fixed search path to use stringified table
        instead of model moniker

0.2_03 Tue Nov 16, 00:10:00 2004
    - fixed bug in link template
    - added dhandler back - it's still needed, though simplified

0.2_02 Mon Nov 15, 23:20:00 2004
    - fixed search path - table-specific paths were not being searched by
        components other than the main request component
    - removed the dhandler, no longer needed with fixed search path
    - added set_mason_comp_roots method, which allows subclasses to fix up
        their own component root search path
    - minor tweaks to a couple of factory templates

0.2_01 Thu Nov 12, 17:30:00 2004
    - initial developer release with Maypole 2 support
    - completely rebuilt to support Maypole 2
    - new templates and css brought in from Maypole 2
    - uses Module::Build for distribution
    - now includes a view component (MasonX::Maypole::View)
    - templates search path matches that in the main Maypole documentation
    - ditto for path parsing

0.04  Thu Aug 19 11:05:00 2004
    - POD updates

0.03  Fri Aug  6 09:45:00 2004
    - minor additions to maypole.css
    - completed XHTMLisation of templates

0.02  Sun Aug  1 23:29:58 2004
    - added missing CSS file, fixed typos in POD

0.01  Sun Aug  1 15:30:53 2004
	- original version; created by h2xs 1.21 with options
		-XAn MasonX::Maypole



INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


COPYRIGHT AND LICENCE

Copyright (C) 2004 David Baird

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for MasonX-Maypole

Change log
==========

0.51 Sat Sept 24, 2005 21:35
    - fixed bug in memoize normalizer function in MasonX::Maypole::View - 
        if running multiple apps, the memoized paths() method would return the same 
        paths for non-model URLs in different apps, which was breaking Mason's 
        dynamic component root system

0.5 Fri Aug 12, 2005 23:55
    - ** INCOMPATIBLE CHANGES **
        - removed MasonX::Maypole::View::error() - calls to the view's error method 
            will now be handled by Maypole::View::Base::error()
        - do *not* set $config->masonx->{comp_root} - your setting will be ignored and 
            over-written
        - added a new (optional) factory_root slot to the config object
        - removed MasonX::Maypole::set_mason_comp_roots()
    
    - requires HTML::Mason 1.30 - for dynamic comp roots
    - requires Maypole 2.10     - for Maypole::Application support
    
    - management of dynamic component roots has changed to use the new Mason capability
    - added a dependency on Memoize
    - simplified error handling - by removing it. If something wants to 
        die, it just dies. Maypole sorts out catching exceptions, somehow.
    - added MasonX::Maypole::View::paths(), which overrides Maypole::View::Base::paths. 
        This method builds the list of component roots for each request. In fact, it 
        wraps the new MasonX::Maypole::_paths() method. 

0.423 Tues Jul 19, 2005 2:30
    - added pod-coverage.t

0.422 Sun  Jun 26, 13:40:00 2005
    - um, fixed syntax error in send_output()

0.421 Sun  Jun 26, 13:30:00 2005
    - fixed syntax error in send_output()

0.42  Fri  Jun 24, 10:00:00 2005
    - now supports pre- and post- 1.26 HTML::Mason
    - changed the label for model comp roots in send_output()
    - send_output() now returns a status code (although Maypole doesn't check it)

0.4_1 Thur Jun 23, 00:30:00 2005
    - added support for HTML::Mason 1.29_1 (and broke support 
        for all previous releases)

0.34 Tues May 31, 20:15:00 2005
    - moved maypole.css to correct directory

0.33 Sun May 22, 23:45:00 2005
    - removed MasonX::Maypole::Application and replaced it with a link 
        to beerdb.riverside-cms.co.uk, where a patched Maypole::Application
        can be downloaded.

0.32 Sun May 22, 23:15:00 2005
    - view class now defaults to MasonX::Maypole::View

0.31 Wed May 18, 19:46:00 2005
    - fixed the package declaration in MasonX::Maypole::Application

0.3  Mon May 16, 17:35:00 2005
    - Fixed important bug where MasonX::Maypole::send_output() wasn't transferring 
        headers_out from the Maypole request to the Apache request
    - Ugh. The bug in 0.223 isn't fixed, just the warning.
    - Removed extra warnings from View.pm
    - Added MasonX::Maypole::Application until MP::App gets patched.

0.223  Mon May 16, 17:30:00 2005
    - it seems that Apache's register_cleanup method should not return undef. 
        Who knew? Anyway, now returning 1 in the register_cleanup call in 
        MasonX::Maypole::View::template(). This fixes an obscure bug where 
        template vars were not getting cleaned up at the end of each request. 
        It also fixes an annoying warning about using undefined variables, but 
        without any file name or line number to find the damned thing. 

0.222  Wed Apr 13, 13:30:00 2005
    - added init anf flags sections to maypole.css to prevent Mason from 
        wrapping the css content, and to ensure the file is served with 
        the correct content type.

0.221  Fri Apr  8, 15:00:00 2005
    - slight modification to frontpage template to remove extra title
    - added information about beerdb demo at http://beerdb.riverside-cms.co.uk

0.220 Thur Apr  7, 03:20:00 2005
    - fixed a couple of bugs in maybe_link_view template
    - edited header template to place maypole.css in /css
    - updated example httpd.conf and BeerDB.pm

0.219  Sun Feb 20, 18:03:00 2005
    - changed send_output() to add a component root based on the model moniker, rather 
        than the table name. In most cases these will be identical. 
    - fixed a bug where all custom component roots were compiled in the same subdirectory 
        of the Mason data dir, thus breaking the ability of different versions of the 
        same template to be called for different tables (mreece). Can't figure out why 
        this never affected me, since it's pretty fundamental to the whole thing. 

0.218  Tues Feb 15, 15:40:00 2005
    - fixed bug in search_form template ($element->{tag} should be $element->tag)

0.217  Fri  Feb 11, 11:55:00 2005
    - fixed links in frontpage template 

0.216  Thur Feb 10, 22:30:00 2005
    - added Makefile.PL (was just Build.PL before)
    - fixed typo in synopsis (mreece)
    - fixed bug caused by not initialising the masonx config slot (mreece)

0.215  Wed Dec 15, 15:00:00 2004
    - fixed bug in pager template

0.213  Thu Dec 9, 01:15:00 2004
    - applied patch to view template, splitting part of it out into the
        new view_item template

0.212  Wed Dec 8, 13:15:00 2004
    - applied patch to button template, to check if action is_public

0.211  Tue Dec 7, 17:30:00 2004
    - applied patch to maybe_link_view template (the UNIVERSAL::isa check)

0.21   Tue Dec 7, 14:00:00 2004
    - removed some bad POD from templates section
    - removed developer status from version string - the module seems to be
      working well on a development site

0.2_05 Sat Nov 20, 22:32:00 2004
    - fixed a bug in send_output - wasn't returning if delegating to SUPER

0.2_04 Tue Nov 16, 01:05:00 2004
    - removed debugging messages, fixed search path to use stringified table
        instead of model moniker

0.2_03 Tue Nov 16, 00:10:00 2004
    - fixed bug in link template
    - added dhandler back - it's still needed, though simplified

0.2_02 Mon Nov 15, 23:20:00 2004
    - fixed search path - table-specific paths were not being searched by

lib/MasonX/Maypole/View.pm
package MasonX::Maypole::View;
use warnings;
use strict;

use Maypole::Constants;
use Memoize;
use Symbol 'qualify_to_ref';

use base 'Maypole::View::Base';

=head1 NAME

MasonX::Maypole::View - Mason view subclass for MasonX::Maypole + Maypole 2

=head1 SYNOPSIS

See L<MasonX::Maypole|MasonX::Maypole>.

=head1 METHODS

=over

=item template

Loads the Maypole template vars into Mason components' namespace.

=cut

sub template 
{
    my ( $self, $r ) = @_;
    
    my $pkg = $r->config->masonx->{in_package};

    my %vars = $self->vars( $r );

    warn __PACKAGE__ . " - template vars maybe not getting cleaned up" if $r->debug > 1;
    
    foreach my $varname ( keys %vars )
    {
        my $export = qualify_to_ref( $varname, $pkg );
        *$export = \$vars{ $varname };

        # this does _not_ seem to be cleaning up always?
        $r->ar->register_cleanup( sub { undef *$export; 1 } );
    }

    return OK;
}

=item paths

Builds the list of component roots in the correct order for Mason to search:

    table-specific      - if path exists
    custom              - if path exists
    template root
    factory 

=cut

memoize( 'paths', NORMALIZER => sub { shift; my $r = shift; $r->model_class || ref $r } );

# this returns config info, so should be in the controller
sub paths 
{
    my ( $self, $r ) = @_;
    
    return $r->_paths;
}

1;



lib/MasonX/Maypole/Application.pm
package MasonX::Maypole::Application;

use strict;
use warnings;
use UNIVERSAL::require;
use Maypole;
use Maypole::Config;

our @ISA;
our $VERSION = '2.09';

sub import {
    my ( $class, @plugins ) = @_;
    my $caller = caller(0);
    
    my $frontend = 'Apache::MVC' if $ENV{MOD_PERL};
    
    my $masonx;
    if ( grep { /^MasonX$/ } @plugins )
    {
        $masonx++;
        @plugins = grep { ! /^MasonX$/ } @plugins;
        $frontend = 'MasonX::Maypole';
    }
    
    $frontend ||= 'CGI::Maypole';
    
    $frontend->require or die "Loading $frontend frontend failed: $@";
    push @ISA, $frontend;

    my $autosetup=0;
    my @plugin_modules;
    {
        foreach (@plugins) {
            if    (/^\-Setup$/) { $autosetup++; }
            elsif (/^\-Debug(\d*)$/) {
                my $d = $1 || 1;
                no strict 'refs';
                *{"$caller\::debug"} = sub { $d };
                warn "Debugging (level $d) enabled for $caller";
            }
            elsif (/^-.*$/) { warn "Unknown flag: $_" }
            else {
                my $plugin = "Maypole::Plugin::$_";
                if ($plugin->require) {
                    push @plugin_modules, "Maypole::Plugin::$_";
                    warn "Loaded plugin: $plugin for $caller"
                        if $caller->can('debug') && $caller->debug;
                } else {
                    die qq(Loading plugin "$plugin" for $caller failed: )
                        . $UNIVERSAL::require::ERROR;
                }
            }
        }
    }
    no strict 'refs';
    push @{"${caller}::ISA"}, @plugin_modules, $class;
    $caller->config(Maypole::Config->new);
    $caller->config->masonx({}) if $masonx;
    $caller->setup() if $autosetup;
}

1;

=head1 NAME

Maypole::Application - Universal Maypole Frontend

=head1 SYNOPSIS

    use Maypole::Application;

    use Maypole::Application qw(Config::YAML);

    use Maypole::Application qw(-Debug Config::YAML -Setup);

    use Maypole::Application qw(Config::YAML Loader -Setup -Debug);

    use Maypole::Application qw(-Debug2 MasonX AutoUntaint);

=head1 DESCRIPTION

This is a universal frontend for mod_perl1, mod_perl2, HTML::Mason and CGI.

You can omit the Maypole::Plugin:: prefix from plugins.
So Maypole::Plugin::Config::YAML becomes Config::YAML.

    use Maypole::Application qw(Config::YAML);

You can also set special flags like -Setup and -Debug.

    use Maypole::Application qw(-Debug Config::YAML -Setup);

The position of plugins and flags in the chain is important,
because they are loaded/executed in the same order they appear.

=head2 -Setup

    use Maypole::Application qw(-Setup);

is equivalent to

    use Maypole::Application;
    MyApp->setup;

Note that no options are passed to C<setup()>. You must ensure that the
required model config parameters are set in C<MyApp-E<gt>config>. See
L<Maypole::Config> for more information.

=head2 -Debug

    use Maypole::Application qw(-Debug);

is equivalent to

    use Maypole::Application;
    sub debug { 1 }

You can specify a higher debug level by saying C<-Debug2> etc. 

=head1 AUTHOR

Sebastian Riedel, C<sri@oook.de>
Idea by Marcus Ramberg, C<marcus@thefeed.no>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut


lib/MasonX/Maypole.pm
package MasonX::Maypole;
use warnings;
use strict;
use Carp;

use HTML::Mason 1.30; # for dynamic comp roots
use Maypole 2.10;     # for Maypole::Application support

# need to get rid of this, and a couple of bits and pieces, to allow CGI mode
use base 'Apache::MVC';

our $VERSION = 0.51;

Maypole::Config->mk_accessors( qw( masonx factory_root ) );

__PACKAGE__->mk_classdata( 'mason_ah' );

=head1 NAME

MasonX::Maypole - use Mason as the frontend and view for Maypole version 2

=head1 SYNOPSIS

    package BeerDB;
    use warnings;
    use strict;

    use Class::DBI::Loader::Relationship;

    use Maypole::Application qw( -Debug2 MasonX AutoUntaint );

    BeerDB->setup( 'dbi:mysql:beerdb', 'username', 'password' );

    BeerDB->config->{template_root}  = '/home/beerdb/www/www/htdocs';
    BeerDB->config->{uri_base}       = '/';
    BeerDB->config->{rows_per_page}  = 10;
    BeerDB->config->{display_tables} = [ qw( beer brewery pub style ) ];
    BeerDB->config->{application_name} = 'The Beer Database';

    BeerDB->config->masonx->{comp_root}  = [ [ factory => '/var/www/maypole/factory' ] ];
    BeerDB->config->masonx->{data_dir}   = '/home/beerdb/www/www/mdata';
    BeerDB->config->masonx->{in_package} = 'BeerDB::TestApp';

    BeerDB->auto_untaint;

    BeerDB->config->{loader}->relationship($_) for (
        'a brewery produces beers',
        'a style defines beers',
        'a pub has beers on handpumps',
        );

    1;

=head1 ** API CHANGES **

Version 0.5 contains major modifications, and changes to error handling. See 
the Changes file for details.

=head1 DESCRIPTION

A frontend and view for Maypole, using Mason.

=head1 EXAMPLES

Example C<BeerDB.pm> and a C<httpd.conf> VirtualHost setup are included in 
the C</doc> directory of the distribution.

A working example of the BeerDB application is at C<http://beerdb.riverside-cms.co.uk>, 
including the C<BeerDB.pm> and C<httpd.conf> used for that site.

=head1 CONFIGURING MASON

Set any parameters for the Mason ApacheHandler in C<<BeerDB->config->{masonx}>>.
This is where to tell Maypole/Mason where the factory templates are stored.

Note that the user the server runs as must have permission to read the files in the
factory templates directory, which also means all directories in the path to the
templates must be readable and executable (i.e. openable). If Mason can't read
these templates, you may get a cryptic 'file doesn't exist' error, but you
will not get a 'not permitted' error.

=head1 Maypole::Application

L<Maypole::Application|Maypole::Application> needs to be patched before it will work 
with MasonX::Maypole. You can download a patched copy from C<http://beerdb.riverside-cms.co.uk>, 
until the required updates are included in the version distributed with L<Maypole>.

=head1 TEMPLATES

This distribution includes Masonized versions of the standard Maypole templates,
plus a dhandler and autohandler. The autohandler simply takes care of adding
a header and footer to every page, while the dhandler loads the template
specified in the Maypole request object.

So if you set the factory comp_root to point at the Maypole factory templates,
the thing should Just Work right out of the box. Except for maypole.css, which 
you will need to copy to the right place on your server. 

=head1 METHODS

=over

=item init

This method is called by Maypole while processing the first request the server
receives. Probably better under mod_perl to call this explicitly at the end of
your setup code (C<BeerDB-E<gt>init>) to share memory among Apache children.
Sets up the Mason ApacheHandler, including the search path behaviour.

=cut

# This only gets called once. Mason's path searching mechanism replaces
# get_template_root and Maypole::View::Base::paths.
sub init 
{
    my ( $r ) = @_;

    my $mason_cfg = $r->config->masonx;
    
    $mason_cfg->{comp_root} = [ $r->_paths ];
    
    $mason_cfg->{decline_dirs} ||= 0;
    $mason_cfg->{in_package}   ||= 'HTML::Mason::Commands';
    $mason_cfg->{dynamic_comp_root} = 1;
    
    $r->mason_ah( MasonX::Maypole::ApacheHandler->new( %{ $mason_cfg } ) );
    
    $r->config->view || $r->config->view( 'MasonX::Maypole::View' );
        
    # set up the view object
    $r->SUPER::init;
}

# MasonX::Maypole::View::path() wraps this
sub _paths
{
    my ( $r ) = @_; # class during startup, object during requests

    my $root    = $r->config->template_root || $r->get_template_root;
    my $factory = $r->config->factory_root; 
    
    $root = ref $root eq 'ARRAY' ? $root : [ $root ];
    
    my @paths;

    foreach my $path ( @$root ) 
    {
        if ( ref $r ) # false during startup - calling model_class() is a fatal error
        {
            if ( my $model = $r->model_class )
            {
                my $model_path = File::Spec->catdir( $path, $model->moniker );
                
                $model =~ s/::/_/g;
                
                push @paths, [ $model, $model_path ] if -d $model_path;
            }
        }
                
        my $custom = File::Spec->catdir( $path, 'custom'  );
        
        push @paths, [ custom => $custom ] if -d $custom;
        
        push @paths, [ default => $path ];
        
        my $this_factory = File::Spec->catdir( $path, 'factory' );
        push @paths, [ factory => $this_factory ] if -d $this_factory;
    }
    
    push @paths, [ factory => $factory ] if $factory;
    
    return @paths;
}

=item parse_args

Uses Mason to extract the request arguments from the request.

=cut

# override the method in Apache::MVC
sub parse_args 
{
    my ( $r ) = @_;

    # set and return request args in Mason request object
    my $args = $r->mason_ah->request_args( $r->ar );

    $r->{params} = $args;
    $r->{query}  = $args;
}

=item send_output

Template variables have already been exported to Mason components namespace
in C<MasonX::Maypole::View::template>. This method now runs the Mason C<exec>
phase to generate and send output.

=cut

sub send_output 
{
    my ( $r ) = @_;
    
    #
    # set up the Mason request object
    #
    
    # if there was an error, there may already be a report in the output slot
    die $r->output if $r->output;

    my $m = $r->mason_ah->prepare_request( $r->ar );
    
    unless ( ref $m )
    {
        warn "prepare_request returned this: [$m] instead of a Mason request object";
        return $m;
    }
    
    #
    # set headers (Mason will set content-length and call send_http_header)
    #
    $r->ar->content_type(
          $r->content_type =~ m/^text/
        ? $r->content_type . "; charset=" . $r->document_encoding
        : $r->content_type
    );
    
    foreach ( $r->headers_out->field_names ) 
    {
        next if /^Content-(Type|Length)/;
        $r->{ar}->headers_out->set( $_ => $r->headers_out->get( $_ ) );
    }
    
    #
    # set dynamic comp roots
    #
    $m->interp->comp_root( [ $r->view_object->paths( $r ) ] );

    if ( $r->debug > 1 )
    {
        Data::Dumper->require or die "Failed to load Data::Dumper: $@";
        warn "Comp roots: " . Data::Dumper::Dumper( $m->interp->comp_root );
    }
    
    # now generate and send output
    my $status = $m->exec;
    
    # Maypole doesn't actually check this status, but for the sake of good form:
    return $status;
}

=item get_template_root

Returns C<template_root> from the config.

This varies from L<Apache::MVC|Apache::MVC>, which concatenates
document_root and location from the Apache request server config.

=cut

sub get_template_root { $_[0]->config->template_root }

=item get_request

Replaces C<Apache::MVC::get_request>, using C<Apache::Request::instance()> instead 
of C<Apache::Request::new()> to obtain the APR object. Calling C<new> means Mason 
and Maypole have different APR objects, and the Mason one doesn't have any POST 
data.

=cut

sub get_request 
{
    my ( $self, $r ) = @_;
    
    $self->{ar} = Apache::Request->instance( $r ); 
}

{
    # copied from MasonX::WebApp
    package MasonX::Maypole::ApacheHandler;
    use base 'HTML::Mason::ApacheHandler';

    sub request_args
    {
        my ( $self, $apr ) = @_;

        return $apr->pnotes('__request_args__') if $apr->pnotes('__request_args__');

        my $args = ($self->SUPER::request_args($apr))[0] || {};

        $apr->pnotes( __request_args__ => $args );

        return $args;
    }
}

=back

=head1 AUTHOR

David Baird, C<< <cpan@riverside-cms.co.uk> >>

=head1 TODO

Currently hard-coded to use Apache/mod_perl. Shouldn't be too hard to use CGI
instead.

=head1 BUGS

Please report any bugs or feature requests to
C<bug-masonx-maypole2@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.  I will be notified, and then you'll automatically
be notified of progress on your bug as I make changes.

=head1 TESTS

There are none. The module loads Mason::ApacheHandler, which causes compile
time errors unless loaded within mod_perl.

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2004 David Baird, All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of MasonX::Maypole2
